exposed-dao/src/main/kotlin/org/jetbrains/exposed/dao/EntityCache.kt

```diff
@@ -5,25 +5,29 @@ import org.jetbrains.exposed.dao.id.IdTable
 import org.jetbrains.exposed.sql.*
 import org.jetbrains.exposed.sql.transactions.transactionScope
 import java.util.*
-import java.util.concurrent.CopyOnWriteArrayList
+import kotlin.collections.HashMap
 
 val Transaction.entityCache : EntityCache by transactionScope { EntityCache(this) }
 
 @Suppress("UNCHECKED_CAST")
 class EntityCache(private val transaction: Transaction) {
     internal var flushingEntities by transactionScope { false }
-    val data = LinkedHashMap<IdTable<*>, MutableMap<Any, Entity<*>>>()
-    val inserts = LinkedHashMap<IdTable<*>, MutableList<Entity<*>>>()
-    val referrers = HashMap<EntityID<*>, MutableMap<Column<*>, SizedIterable<*>>>()
+    internal val data = LinkedHashMap<IdTable<*>, MutableMap<Any, Entity<*>>>()
+    internal val inserts = LinkedHashMap<IdTable<*>, MutableSet<Entity<*>>>()
+    internal val updates = LinkedHashMap<IdTable<*>, MutableSet<Entity<*>>>()
+    internal val referrers = HashMap<Column<*>, MutableMap<EntityID<*>, SizedIterable<*>>>()
```

Comment:
I've flipped the hierarchy of `referrers` to allow bulk operations by column rather than entity.  `referrerEntities` is maintained to determine which entities have which columns.  As noted elsewhere, the functionality of `referrerEntities` could be accomplished by listing the referrer properties, but there is currently no easy way to do that within the library.